report_substitute/README.rst

```diff
@@ -0,0 +1,100 @@
+=================
+Report Substitute
+=================
+
+.. !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+   !! This file is generated by oca-gen-addon-readme !!
+   !! changes will be overwritten.                   !!
+   !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+.. |badge1| image:: https://img.shields.io/badge/maturity-Beta-yellow.png
+    :target: https://odoo-community.org/page/development-status
+    :alt: Beta
+.. |badge2| image:: https://img.shields.io/badge/licence-AGPL--3-blue.png
+    :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+    :alt: License: AGPL-3
+.. |badge3| image:: https://img.shields.io/badge/github-OCA%2Freporting--engine-lightgray.png?logo=github
+    :target: https://github.com/OCA/reporting-engine/tree/12.0/report_substitute
+    :alt: OCA/reporting-engine
+.. |badge4| image:: https://img.shields.io/badge/weblate-Translate%20me-F47D42.png
+    :target: https://translation.odoo-community.org/projects/reporting-engine-12-0/reporting-engine-12-0-report_substitute
+    :alt: Translate me on Weblate
+.. |badge5| image:: https://img.shields.io/badge/runbot-Try%20me-875A7B.png
+    :target: https://runbot.odoo-community.org/runbot/143/12.0
+    :alt: Try me on Runbot
+
+|badge1| |badge2| |badge3| |badge4| |badge5| 
+
+This module allows you to create substitution rules for report actions.
+A typical use case is to replace a standard report by alternative reports
+when some conditions are met. For instance, it allows to configure alternate
+reports for different companies.
```

Comment:
To manage alternative reports for different companies, I find the existing OCA module [ir_actions_report_multi_company](https://github.com/OCA/multi-company/tree/11.0/ir_actions_report_multi_company) much easier for a non-technical user.

I would change this line in the readme: instead of mentioning reports for different companies, I would expect some other use case. A real case with some code examples would be here very appreciated.

Reply:
In standard modules, many report calls are hard-coded by their xml ID. An example, the print quotation action in the sale module.

https://github.com/odoo/odoo/blob/48b3893b84cbd51f880548b38d036db5ebc1e256/addons/sale/models/sale.py#L489-L494

Adding an ir_rule to the report action will not resolve the use case here because this type of action will raise an access error for other companies if a company is set.

the alternative solutions that can be made in this case and that we avoid with this module:
- override the print action to improve the way odoo get the report action.
- override the report template to manage all the layout differences in the same template.

So the idea is to define some substitution rules on the standard report to return the desired one if some condition are met.

Reply:
IMO, with the domain widget used in the UI, the the substitution rules is easy to use for non-technical user